File: src/test/java/fr/marstech/mtlinkspray/controller/ViewSprayControllerTest.java
package fr.marstech.mtlinkspray.controller;

import fr.marstech.mtlinkspray.controller.api.RootApiController;
import fr.marstech.mtlinkspray.controller.view.ViewSprayController;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.result.MockMvcResultMatchers;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
@WebMvcTest(ViewSprayController.class)
class ViewSprayControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Test
    void getSprayPage() throws Exception {
        mockMvc.perform(MockMvcRequestBuilders.get("/spray")).andExpect(MockMvcResultMatchers.status().isOk());
    }

    @Test
    void getLink() throws Exception {
        mockMvc.perform(MockMvcRequestBuilders.post("/spray").param("inputLinkList", "test")).andExpect(MockMvcResultMatchers.status().isOk());
    }

    @Test
    void getSpray() throws Exception {
        mockMvc.perform(MockMvcRequestBuilders.get("/spray/open").param("spray", "test")).andExpect(MockMvcResultMatchers.status().isOk());
    }

    @TestConfiguration
    static class ApiRootControllerTestConfig {
        @Bean(name = "apiRootController")
        public RootApiController apiRootController() {
            RootApiController mock = mock(RootApiController.class);
            when(mock.getVersion()).thenReturn("test-version");
            return mock;
        }
    }
}

File: src/test/kotlin/fr/marstech/mtlinkspray/controller/ViewSprayControllerTest.kt
package fr.marstech.mtlinkspray.controller

import fr.marstech.mtlinkspray.controller.view.ViewSprayController
import fr.marstech.mtlinkspray.service.SprayService
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.web.servlet.ModelAndView
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class ViewSprayControllerTest {

    private val sprayService: SprayService = mock(SprayService::class.java)
    private val controller = ViewSprayController(sprayService)

    @Test
    fun getSprayPageShouldReturnCorrectView() {
        val request = MockHttpServletRequest()
        val result: ModelAndView = controller.getSprayPage(request)

        assertEquals("spray", result.viewName)
        assertNotNull(result.model["headers"])
    }

    @Test
    fun postSprayShouldReturnCorrectViewWithModel() {
        val request = MockHttpServletRequest()
        val result: ModelAndView = controller.getLink("https://example.com", request)

        assertEquals("spray", result.viewName)
        assertNotNull(result.model["linkList"])
        assertNotNull(result.model["linkListText"])
        assertNotNull(result.model["linkSpray"])
    }

    @Test
    fun postSprayShouldReturnEmptyLinkListWhenInputIsBlank() {
        // Given
        val request = MockHttpServletRequest()

        // When
        val result: ModelAndView = controller.getLink("   ", request)

        // Then
        assertEquals("spray", result.viewName)
        val linkList = result.model["linkList"] as? List<*>
        assertNotNull(linkList)
        assertTrue(linkList.isEmpty())
    }

    @Test
    fun getSprayOpenShouldReturnCorrectViewWithModel() {
        val sprayList = listOf("https://example.com")
        val result: ModelAndView = controller.getSpray(sprayList)

        assertEquals("spray", result.viewName)
        assertNotNull(result.model["linkListText"])
        assertNotNull(result.model["spray"])
    }

    @Test
    fun postShortenedSprayShouldReturnCorrectViewWithShortenedLink() {
        // Given
        val request = MockHttpServletRequest()
        val shortenedUrl = "http://localhost/abc123"
        `when`(sprayService.shortenAndSpray(listOf("https://example.com"), request))
            .thenReturn(shortenedUrl)

        // When
        val result: ModelAndView = controller.getShortenedSpray("https://example.com", request)

        // Then
        assertEquals("spray", result.viewName)
        assertNotNull(result.model["linkList"])
        assertNotNull(result.model["linkListText"])
        assertEquals(shortenedUrl, result.model["linkSpray"])
        assertEquals(true, result.model["linkSprayShortened"])
    }
}